File: sdks/typescript/src/ui/emitter.ts
import type {
  AppendEvent,
  EndEvent,
  ErrorEvent,
  PingEvent,
  SetBlockEvent,
  SetMetaEvent,
  StartEvent,
} from "./event.ts";
import { cloneJson, type JsonRecord } from "./json.ts";
import type { BaseBlock, ModelMeta, UIModel } from "./model.ts";

export interface EventOptions {
  eventType?: string;
}

export interface BlockSetOptions extends EventOptions {
  mask?: `block.${string}`;
}

export interface BlockAppendOptions extends EventOptions {
  mask?: `block.${string}`;
}

export interface ErrorOptions {
  traceId?: string;
  detail?: string;
}

export class PatchEmitter {
  private currentOffset: number;

  constructor(startOffset = 0) {
    if (!Number.isInteger(startOffset) || startOffset < 0) {
      throw new Error("startOffset must be a non-negative integer");
    }
    this.currentOffset = startOffset;
  }

  get offset(): number {
    return this.currentOffset;
  }

  start(model: UIModel, seq?: number): StartEvent {
    const nextSeq = seq ?? this.nextOffset();
    if (!Number.isInteger(nextSeq) || nextSeq < this.currentOffset) {
      throw new Error("start seq cannot move the emitter backwards");
    }
    this.currentOffset = nextSeq;
    return { op: "start", seq: nextSeq, model: cloneJson(model) };
  }

  blockSet(block: BaseBlock, options: BlockSetOptions = {}): SetBlockEvent {
    const seq = this.nextOffset();
    const event: SetBlockEvent = options.mask === undefined
      ? { op: "set", seq, block: cloneJson(block) }
      : { op: "set", seq, mask: options.mask, block: cloneJson(block) };
    if (options.eventType !== undefined) event.event_type = options.eventType;
    return event;
  }

  metaSet(
    mask: `meta.${string}`,
    meta: Partial<ModelMeta> & JsonRecord,
    options: EventOptions = {},
  ): SetMetaEvent {
    const event: SetMetaEvent = {
      op: "set",
      seq: this.nextOffset(),
      mask,
      meta: cloneJson(meta),
    };
    if (options.eventType !== undefined) event.event_type = options.eventType;
    return event;
  }

  blockAppend(block: BaseBlock, options: BlockAppendOptions = {}): AppendEvent {
    const event: AppendEvent = {
      op: "append",
      seq: this.nextOffset(),
      mask: options.mask ?? "block.content",
      block: cloneJson(block),
    };
    if (options.eventType !== undefined) event.event_type = options.eventType;
    return event;
  }

  error(code: string, message: string, options: ErrorOptions = {}): ErrorEvent {
    const error: JsonRecord = { code, message };
    if (options.detail !== undefined && options.detail !== message) error.detail = options.detail;
    if (options.traceId !== undefined) error.trace_id = options.traceId;
    return {
      op: "error",
      seq: this.nextOffset(),
      mask: "meta.error",
      meta: { error },
    };
  }

  setStats(stats: JsonRecord): SetMetaEvent {
    return this.metaSet("meta.stats", { stats: cloneJson(stats) });
  }

  ping(): PingEvent {
    return { op: "ping", seq: this.currentOffset, ts: Date.now() };
  }

  end(): EndEvent {
    return { op: "end", seq: this.nextOffset() };
  }

  private nextOffset(): number {
    this.currentOffset += 1;
    return this.currentOffset;
  }
}

export { PatchEmitter as SSEEmitter };

File: sdks/python/src/agentue/emitter.py
"""Ordered AgentUE patch construction."""

import time
from collections.abc import Mapping
from typing import Any

from agentue.event import PatchEvent, PatchOp
from agentue.model import BaseBlock, ReferenceBlock, UIModel


class PatchEmitter:
    """Stateful event builder that manages ``seq`` values."""

    def __init__(self, start_offset: int = 0):
        if start_offset < 0:
            raise ValueError("start_offset must be non-negative")
        self._offset = start_offset

    @property
    def offset(self) -> int:
        return self._offset

    def _next_offset(self) -> int:
        self._offset += 1
        return self._offset

    def start(self, model: UIModel, *, seq: int | None = None) -> str:
        """Create the first event in a delivery stream.

        ``seq`` can identify the highest event covered by a reconstructed
        snapshot. Without it, the emitter advances its current offset.
        """
        if seq is None:
            seq = self._next_offset()
        else:
            if seq < self._offset:
                raise ValueError("start seq cannot move the emitter backwards")
            self._offset = seq

        event = PatchEvent(
            op=PatchOp.START,
            seq=seq,
            model=model.model_dump(mode="json", exclude_none=True, serialize_as_any=True),
        )
        return event.to_json()

    def block_set(
        self, block: BaseBlock | ReferenceBlock, mask: str | None = None, *, event_type: str | None = None
    ) -> str:
        """Create or replace a block, or set one of its fields."""
        payload = block.model_dump(
            mode="json",
            exclude_none=mask is None,
            exclude_unset=mask is not None,
            serialize_as_any=True,
        )
        event = PatchEvent(
            op=PatchOp.SET,
            seq=self._next_offset(),
            mask=mask,
            event_type=event_type,
            block=payload,
        )
        return event.to_json()

    def meta_set(self, mask: str, meta: Mapping[str, Any], *, event_type: str | None = None) -> str:
        """Set a metadata field selected by ``mask``."""
        event = PatchEvent(
            op=PatchOp.SET,
            seq=self._next_offset(),
            mask=mask,
            event_type=event_type,
            meta=dict(meta),
        )
        return event.to_json()

    def block_append(
        self,
        block: BaseBlock,
        mask: str = "block.content",
        *,
        event_type: str | None = None,
    ) -> str:
        """Append a string or list field to a block."""
        event = PatchEvent(
            op=PatchOp.APPEND,
            seq=self._next_offset(),
            mask=mask,
            event_type=event_type,
            block=block.model_dump(mode="json", exclude_none=True, serialize_as_any=True),
        )
        return event.to_json()

    def error(
        self,
        code: str,
        message: str,
        *,
        trace_id: str | None = None,
        detail: str | None = None,
    ) -> str:
        """Set ``meta.error`` and signal a failed delivery."""
        error_payload: dict[str, str] = {"code": code, "message": message}
        if detail is not None and detail != message:
            error_payload["detail"] = detail
        if trace_id is not None:
            error_payload["trace_id"] = trace_id
        event = PatchEvent(
            op=PatchOp.ERROR,
            seq=self._next_offset(),
            mask="meta.error",
            meta={"error": error_payload},
        )
        return event.to_json()

    def set_stats(self, stats: Mapping[str, Any]) -> str:
        """Convenience helper for the common ``meta.stats`` extension."""
        return self.meta_set("meta.stats", {"stats": dict(stats)})

    def ping(self) -> str:
        """Create a heartbeat without advancing ``seq``."""
        return PatchEvent(op=PatchOp.PING, seq=self._offset, ts=int(time.time() * 1000)).to_json()

    def end(self) -> str:
        """Create the final event in a delivery stream."""
        return PatchEvent(op=PatchOp.END, seq=self._next_offset()).to_json()


# Compatibility name for applications migrating from an SSE-specific emitter.
SSEEmitter = PatchEmitter
